components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AbstractApplicationAuthenticator.java

```diff
@@ -352,6 +360,74 @@ protected String getUserStoreAppendedName(String userName) {
         return userName;
     }
 
+    protected org.wso2.carbon.user.core.common.User getUser(AuthenticatedUser authenticatedUser)
+            throws AuthenticationFailedException {
+
+        org.wso2.carbon.user.core.common.User user = null;
+        String tenantDomain = authenticatedUser.getTenantDomain();
+        if (tenantDomain == null) {
+            return null;
+        }
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        try {
+            UserRealm userRealm = FrameworkServiceDataHolder.getInstance().getRealmService()
+                    .getTenantUserRealm(tenantId);
+            if (userRealm != null) {
+                UserStoreManager userStoreManager = userRealm.getUserStoreManager();
+                List<org.wso2.carbon.user.core.common.User> userList
+                        = ((AbstractUserStoreManager) userStoreManager).getUserListWithID(
+                        USERNAME_CLAIM, authenticatedUser.getUserName(), null);
+                if (userList.isEmpty()) {
+                    userList = ((AbstractUserStoreManager) userStoreManager).getUserListWithID(
+                            EMAIL_ADDRESS_CLAIM, authenticatedUser.getUserName(), null);
+                }
+                userList = getValidUsers(userList);
+                if (CollectionUtils.isEmpty(userList)) {
+                    return null;
+                }
+                if (userList.size() > 1) {
+                    throw new AuthenticationFailedException("There are more than one user with the provided username "
+                            + "claim value: " + authenticatedUser.getUserName());
+                }
+                user = userList.get(0);
+            } else {
+                throw new AuthenticationFailedException("Cannot find the user realm for the given tenant: "
+                        + tenantDomain);
+            }
+        } catch (UserStoreException e) {
+            String msg = "Failed to retrieve the user from the user store.";
+            throw new AuthenticationFailedException(msg, e);
```

Comment:
Since there's only one usage of the `msg` variable, let's remove it and add the message straight to the exception.

Reply:
Fixed with https://github.com/wso2/carbon-identity-framework/pull/5103